fix: allow context of EmbeddedViewRef to be changed and avoid mutating context in NgTemplateOutlet

crisbeto opened the pull request:
Currently `NgTemplateOutlet` recreates its view if its template is swapped out or a context object with a different shape is passed in. If an object with the same shape is passed in, we preserve the old view and we mutate the previous object. This mutation of the original object can be undesirable if two objects with the same shape are swapped between two different template outlets.

The current behavior is a result of a limitation in `core` where the `context` of an embedded view is read-only.

These changes are split up into two commits which aim to resolve the issue:
1. The first commit makes `EmbeddedViewRef.context` writeable so that the object doesn't have to be mutated anymore and there didn't seem to be a good reason why it was read-only to begin with.
2. The second commit resolves the context mutation issue by swapping out the context, rather than mutating it, when a new context object comes into `NgTemplateOutlet`.

Fixes #24515.

crisbeto committed:
fix(core): allow EmbeddedViewRef context to be updated

Currently `EmbeddedViewRef.context` is read-only which means that the only way to update
it is to mutate the object which can lead to some undesirable outcomes if the template
and the context are provided by an external consumer (see #24515).

These changes make the property writeable since there doesn't appear to be a specific
reason why it was readonly to begin with.

atscott reviewed (approved):
reviewed-for: fw-core
reviewed-for: public-api
reviewed-for: size-tracking

atscott commented:
[presubmit](http://test/OCL:351178536:BASE:351178559:1610388418850:c2aa9847)

atscott commented:
FYI - it looks like there might be a real failure in g3 due to this change. I haven't gotten a chance to investigate yet, but the bit of code in the template that seems to be broken is 
```
        <ng-template [ngTemplateOutlet]="createFormTemplate"
                     [ngTemplateOutletContext]="{$implicit: (sourceBackup$ | async)}">
        </ng-template>
        <ng-template #createFormTemplate let-sourceBackup>
          <form [formGroup]="createForm"
                (ngSubmit)="create(sourceBackup)">
        </ng-template>
```
and then the failure is `Cannot read property 'name' of undefined` from the `create` method trying to read `sourceBackup`, which must be undefined. I won't get to investigating this until tomorrow, but maybe that's enough info for you to get started before then.

atscott commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -64,44 +62,9 @@ export class NgTemplateOutlet implements OnChanges {
+    } else if (
+        this._viewRef && changes['ngTemplateOutletContext'] && this.ngTemplateOutletContext) {
+      this._viewRef.context = this.ngTemplateOutletContext;
     }
```
@crisbeto - After investigating the presubmit failure and reviewing the change again, I think this is at least going to have to be marked as a breaking change. The behavior in this PR departs somewhat significantly from what the old behavior was. The old behavior only updated new or changed properties on the existing context and left the others unchanged. I don't know if there was a good reason for doing this. We might need to discuss this again in the next FW sync or standup.

crisbeto replied:
Agreed, let's discuss it. IMO the old behavior is somewhat unexpected, but I see how it can be breaking.

crisbeto committed:
fix(common): avoid mutating context object in NgTemplateOutlet

Currently `NgTemplateOutlet` recreates its view if its template is swapped out or a context
object with a different shape is passed in. If an object with the same shape is passed in,
we preserve the old view and we mutate the previous object. This mutation of the original
object can be undesirable if two objects with the same shape are swapped between two
different template outlets.

The current behavior is a result of a limitation in `core` where the `context` of an embedded
view is read-only, however a previous commit made it writeable.

These changes resolve the context mutation issue and clean up a bunch of unnecessary
logic from `NgTemplateOutlet` by taking advantage of the earlier change.

Fixes #24515.

atscott commented:
Hey Kristiyan, I think we're going to have to land a change to Ivy before we can merge this. Here's a test that fails in Ivy only:
```

  fit('has latest context information in listeners in a template', () => {
    const eventLog: string[] = [];
    @Component({
      template: `
        <ng-template #template let-name="name">
          <button (click)="log(name)"></button>
        </ng-template>
        <ng-template [ngTemplateOutlet]="template" [ngTemplateOutletContext]="context"></ng-template>
       `
    })
    class TestCmp {
      context = {name: 'one'};

      log(name: string) {
        eventLog.push(name);
      }
    }

    const fixture =
        TestBed.configureTestingModule({declarations: [TestCmp]}).createComponent(TestCmp);
    fixture.detectChanges();
    const logButton = fixture.nativeElement.querySelector('button');

    logButton.click();
    expect(eventLog).toEqual(['one']);

    fixture.componentInstance.context = {name: 'two'};
    fixture.detectChanges();
    logButton.click();
    expect(eventLog).toEqual(['one', 'two']);
  });
```

The issue is that the generated function for the template looks something like
```
function TestCmp_ng_template_1_Template(rf, ctx) {
  if (rf & 1) {
    ...
    i0.ɵɵlistener("ngSubmit", function InstanceCreateForm_ng_template_8_Template_button_click_0_listener() {
      ...
      var name_r7 = ctx.$implicit;
      ...
    });
  } else if (rf & 2) {
    ...
  }
}
```
So the listener closure is retaining a reference to the original `ctx` at creation time.

crisbeto commented:
Thank you for investigating it @atscott. I've submitted https://github.com/angular/angular/pull/40833 which should resolve the issue.

crisbeto commented:
@atscott I've rebased in order to get the changes from #40833 and I've added your test. Marking for presubmit again.

zarend commented:
[presubmit](http://test/OCL:358195245:BASE:358195273:1613669143868:e7f5603a)

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

petebacondarwin commented on packages/core/src/linker/view_ref.ts:
```diff
@@ -93,7 +93,7 @@ export abstract class EmbeddedViewRef<C> extends ViewRef {
    * The context for this view, inherited from the anchor element.
    */
-  abstract get context(): C;
+  abstract context: C;
```
Strictly this is a breaking change in the case that this class `EmbeddedViewRef` was extended. See [playground](https://www.typescriptlang.org/play?#code/IYIwzgLgTsDGEAJYBthjAgQmgpgg3gFAIIBmA9uQFwIB2ArgLYg5QDchxS5tkU98clAAUFanSYsoASgJcSEABYBLMADoxCALxlKHEgF9CRwijQYAwiuQATBDgAeEHLRsZsYPERIBzHIjFhWXwEKH96KFoEMHoAB1YNPQQjElgePgEIISC5EhIY+JEAJgBGAAZpfWTjQiA).

crisbeto replied:
Do we consider this as something that's meant to be extended, considering that the framework itself constructs it? I believe we've had cases in the past where we've said that constructor changes (also something that would be breaking for extending consumers) aren't breaking.

petebacondarwin replied:
I'm not sure but we could avoid the breaking change by just adding a setter, rather than making it a property.

crisbeto replied:
I wanted to avoid getters and setters, because they generate more code in ES5.

petebacondarwin replied:
And actually adding an abstract setter would also be a breaking change!

atscott replied:
Unless a class is specifically noted that it’s meant to be extended, we don’t consider breaking extenders a breaking change. See https://github.com/angular/angular/pull/32057#issuecomment-574412424

petebacondarwin replied:
So if we are cool that no one outside the framework is going to extend this then I am cool with it

petebacondarwin reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: size-tracking
